tools/slayerd/slayerd.8
.TH SLAYERD 8 "April 19, 2007"
.SH NAME
slayerd \- trim errant user processes
.SH SYNOPSIS
.B slayerd
[-vhf]
.SH DESCRIPTION
.B slayerd
is a daemon which monitors the system for users using too much memory.
If the size of all of a user's processes exceed the maximum limit, slayerd
will kill the user's processes in decreasing order of size, until the
user's memory use drops below the threshold.
.SH OPTIONS
.TP
.B \-h
Show summary of options.
.TP
.B \-v
Show program version.
.SH CONFIGURATION
.B slayerd
reads its configuration from \fI/etc/slayerd/slayerd.conf\fP.  The following
options may be set in this file:
.TP
.B \fBlimit\fP \fInum\fP 
(required)
The maximum amount of memory (in megabytes) that a user may use before their
processes are killed.
.TP
.B \fBthresh\fP \fInum\fP
(required)
When a user exceeds their memory limit, their processes will be killed until
they are using less than this amount of memory (in megabytes).
.TP
.B \fBdelay\fP \fIsecs\fP
(required)
The time in seconds to sleep between each check.
.TP
.B \fBexempt\fP \fIuser\fP
A user who will be exempt from monitoring.  \fIroot\fP is always exempt.  This
option may be specified multiple times.
.TP
.B \fBminuid\fP \fIuid\fP
All users whose user id is less than \fIuid\fP will be exempt.
.TP
.B \fBsendmail\fP \fIcommand\fP
The path to the sendmail command that will be used to send mail.
.TP
.B \fBpidfile\fP \fIpath\fP
Pathname to write pid file to on startup.

.SH MAIL MESSAGE
When a user's processes are killed,
.B slayerd
will send them a mail explaining what happened.  The contents of this message
are in \fI/etc/slayerd/mailmessage\fP.  The following substition variables
may be used in the message:
.TP
.B \fB%hostname%\fP
The hostname of the machine.
.TP
.B \fB%user%\fP
The username of the user whose processes were killed.
.TP
.B \fB%limit%\fP
The value of the \fIlimit\fP option in the configuration.
.TP
.B \fB%thresh%\fP
The value of the \fIthresh\fP option in the configuration.
.TP
.B \fB%current%\fP
The amount of memory used by the user after their processes were killed.
.TP
.B \fB%processes%\fP
The list of processes killed, one per line.

.SH AUTHOR
slayerd was written by River Tarnell <river@attenuate.org>.


wikitex/NEWS









                  WikiTeX 1.1 BETA 1 NEWS

          Peter Danenberg <pcd at wikitex dot org>


Sun Feb 12 16:21:57 PST 2006    Peter Danenberg

          WikiTeX's  security  model  has been reimplemented
     from the ground up:  We've  doffed  our  dependence  on
     blacklists,  and  instead  rely  on  a  combination  of
     ulimit, quotas, sudo; and the native facilites of LaTeX
     and LilyPond for forbidding script execution and direc-
     tory ascent.

          In addition, the `src' and `data' parameters allow
     you to reference uploads directly in your script; exam-
     ples:

          <svg src="tiger.svg"></svg>

     to hide ugly code; or to plot data:

          <plot data="plot.dat">plot "plot.dat" using 1:3</plot>

     Files thus sourced forgo processing by the template.

Thu Mar 10 01:26:39 PST 2005  Peter Danenberg

          Skak by Torben Hoffmann has  replaced  Piet  Tute-
     laers'  package  as the <chess> renderer; Skak is back-
     wards compatible and allows, amongst other things:

          o    SAN and PGN notation

          o    large boards (with \largeboard)

          o    move arrows (with \printarrow)

     see the docs for details.

Tue Mar  8 16:53:39 PST 2005  Peter Danenberg

          <schem> added;  id  est,  schematic  capture  with
     gschem (geda.seul.org).  PPCH has been deprecated; XyM-
     TeX, meanwhile, has taken its place as the <chem>  ren-
     derer.

          Effort  is  underway  to replace our ad-hoc black-
     lists with chroot-based security; adieu,  "directa  non
     grata!"




                      12 February 2006





                             -2-


          Lastly,  verbose  errors  now declare not only the
     whence, but the why of defects.

Sun Feb  6 14:21:48 PST 2005  Peter Danenberg

          We've disassociated our versioning from MediaWiki,
     and  started  with  a  pristine 1.0; so that, should we
     bundle in the future, our release cycles  can  be  dis-
     tinct.

Sat Jan 29 00:34:09 PST 2005  Peter Danenberg

          Changes almost too numerous to enumerate; but some
     of the more exciting:

     o    Gnuplot and Graphviz: <plot> and <graph>,  respec-
          tively

     o    Multi-page output for music, go and chess

     o    More detailed, helpful error messages

          The  installation process, to boot, is less error-
     prone; no Python:  but  the  bash  interface  has  been
     refactored.

Tue Jan 25 11:23:48 PST 2005  Peter Danenberg

          This  candidate  both maintains by tightening down
     our blacklist (thanks, Johannes Berg); and innovates by
     introducing Jan-Aoke Larsson's dvipng.  dvipng replaces
     dvi2bitmap; resulting in cleaner, faster renders.

          <math></math> becomes <amsmath></amsmath>  in  the
     stand-alone  version,  to avoid conflicting with texvc;
     in pre-packaged WikiTeX, it replaces texvc.

          I'd also like to  un-cobweb  that  bash  interface
     with Python before we go live.

Mon Oct  4 01:56:23 PDT 2004  Peter Danenberg

          In  order  to  plumb their several virtues' depth,
     the Batik toolkit  ("heavy  SVG"),  will  be  available
     alongside ImageMagick; its more efficient kin.

          Batik  may  be  invoked  by  <batik></batik>;  and
     <svg></svg>, ImageMagick.

Sun Oct  3 17:38:03 PDT 2004  Peter Danenberg

          SVG is finally here; and we opted for Batik's  SVG
     kit (xml.apache.org/batik) which, though very full-fea-
     tured, suffers from a Java-like retardation.   It  were



                      12 February 2006





                             -3-


     mere  oeconomy  that  ever  importuned  us to return to
     ImageMagick's SVG subset.

Mon Sep  6 03:20:26 PDT 2004  Peter Danenberg

          With the introduction of XML-like  tags,  we  have
     cleared  a major hurdle towards integration with Media-
     Wiki; next on the list, parameters of the form:

          <music tempo="80"></music>

     WikiTeX' vulnerability to attack  must  also  be  thor-
     oughly sounded.

Sun May 30 00:00:00 PDT 2004  Peter Danenberg

          We're   refactoring  WikiTeX from the ground up to
     implement tags of the form:

          <class [parm: value]>content</class>

     and likely of an  external  parser;  where  a  lex/yacc
     solution  might  be attractive.  (Finite state recogni-
     tion is probably what  we're  looking   for,   however,
     rather   than   look-ahead  left-to-right parsing; i.e.
     top-down recursive descent.)

          If  WikiTeX'  sum scope, on the other hand, resem-
     bles a glue-script with  regular-expression  grinding,*
     we  might  not  need  to  look  past  Perl; PHP, in any
     case, just passes as lexer/parser: but we'd like  some-
     thing  more robust and, above all, to separate applica-
     tion from interface.


















-----------
  * Raymond,  Eric  Steven.   The Art of Unix Pro-
gramming (faqs.org/docs/artu).



                      12 February 2006





                             -4-


WikiTeX: expansible LaTeX module for MediaWiki
Copyright (C) 2004-6  Peter Danenberg

     WikiTeX is licensed under the Artistic License 2.0;  to
view a copy of this license, see COPYING or visit:

     http://dev.perl.org/perl6/rfc/346.html


















































                      12 February 2006




tools/slayerd/slayerd.cc
/* Copyright (c) 2007-2009 River Tarnell <river@loreley.flyingparchment.org.uk>. */
/*
 * Permission is granted to anyone to use this software for any purpose,
 * including commercial applications, and to alter it and redistribute it
 * freely. This software is provided 'as-is', without any express or implied
 * warranty.
 */

/* $Id$ */

/*
 * slayerd: monitor user activity and regulate users using too much RAM.
 */

#include	<sys/wait.h>
#include	<sys/mman.h>
#include	<syslog.h>

#include	<cerrno>
#include	<cstring>
#include	<csignal>
#include	<signal.h>
#include	<stdio.h>
#include	<iostream>
#include	<map>
#include	<fstream>
#include	<boost/format.hpp>
#include	<boost/lexical_cast.hpp>

#include	"process.h"
#include	"config.h"
#include	"util.h"

#if defined(__sun) && defined(__SVR4)
extern "C" int daemon(int, int);
#endif

namespace {
	std::string CONFFILE = "/etc/slayerd/slayerd.conf";
}

struct user {
	user() : uid(-1), rss(0) {}

	uid_t uid;
	std::size_t rss;
	std::vector<process::pointer> processes;
};

void run_once();
void check_user(user &u);

void
version(void) {
	std::cerr << "slayerd " << VERSION << "\n";
	std::cerr << "Copyright (C) 2007, River Tarnell <river@attenuate.org>.\n";
}

void
usage(void) {
	std::cerr << "usage: slayerd [-fvh] [-c <conf>]\n"
;
}

extern "C" void
rmpidfile(void)
{
	unlink(config.pidfile.c_str());
}

extern "C" void
do_signal(int sig)
{
	rmpidfile();
	_exit(0);
}

bool
do_pidfile(void)
{
	{
		std::ifstream pf(config.pidfile.c_str());
		if (pf) {
			std::string pid;
			std::getline(pf, pid);
			std::cerr << boost::format("slayerd already running (pid %d) or stale pidfile %s\n")
				% pid % config.pidfile;
			return false;
		}
	}

	std::ofstream pf(config.pidfile.c_str());
	if (!pf) {
		std::cerr << boost::format("cannot open pidfile %s: %s\n")
			% config.pidfile % std::strerror(errno);
		return false;
	}

	pf << getpid() << '\n';
	std::atexit(rmpidfile);
	std::signal(SIGINT, do_signal);
	std::signal(SIGTERM, do_signal);
	return true;
}

int
main(int argc, char **argv)
{
	int fflag = 0;
	int c;

	config.exempt.insert(0);	/* root is always exempt */

	while ((c = getopt(argc, argv, "fvhc:D")) != -1) {
		switch (c) {
			case 'f':
				fflag++;
				break;

			case 'c':
				CONFFILE = optarg;
				break;

			case 'v':
				version();
				return 0;

			case 'h':
				version();
				usage();
				return 0;

			case 'D':
				config.debug = true;
				fflag++;
				break;

			default:
				version();
				usage();
				return 1;
		}
	}
	argc -= optind;
	argv += optind;

	if (!config.debug)
		openlog("slayerd", LOG_PID, LOG_DAEMON);

	if (!configure(CONFFILE))
		return 1;

	if (config.limit == 0 || config.thresh == 0) {
		std::cerr << "invalid limit/threshold\n";
		return 1;
	}

	if (!fflag && daemon(0, 0) == -1) {
		std::cerr << boost::format("cannot daemonise: %s\n") % std::strerror(errno);
		return 1;
	}

	if (!fflag && !do_pidfile())
		return 1;

	if (mlockall(MCL_CURRENT | MCL_FUTURE) == -1) 
		log(str(boost::format("warning: cannot lock memory: %s") % std::strerror(errno)));

	log(str(boost::format("delay: %d, limit: %dM, threshold: %dM")
		% config.delay % (config.limit / 1024 / 1024) % (config.thresh / 1024 / 1024)));

	for (;;) {
		run_once();
		sleep(config.delay);
	}
}

struct process_sort_rss {
	bool operator() (process::pointer a, process::pointer b) const {
		return a->rss() > b->rss();
	}
};

void
run_once()
{
	std::vector<process::pointer> processes(enumerate_processes());

	/*
	 * Aggregate the processes by user.
	 */
	std::vector<user> users;
	for (std::size_t i = 0, end = processes.size(); i < end; ++i) {
		process::pointer p = processes[i];
		user *u = 0;

		for (std::size_t ui = 0, uend = users.size(); ui != uend; ++ui)
			if (users[ui].uid == p->uid()) {
				u = &users[ui];
				break;
			}

		if (u == 0) {
			std::size_t n = users.size();
			users.resize(n + 1);
			users[n].uid = p->uid();
			u = &users[n];
		}

		u->rss += p->rss();
		u->processes.push_back(p);
	}

	for (std::size_t i = 0, end = users.size(); i < end; ++i) {
		check_user(users[i]);
	}
}

void
check_user(user &u)
{
	std::size_t bytes = u.rss;

	if (config.exempt.find(u.uid) != config.exempt.end())
		return;

	if (config.minuid >= 0 && u.uid < config.minuid)
		return;

	if (bytes < config.limit)
		return;

	std::string uname = username(u.uid);
	std::string process_list;

	log(str(boost::format("user \"%s\" is using %dM, over configured limit %dM")
				% uname
				% (bytes / 1024 / 1024)
				% (config.limit / 1024 / 1024)));

	std::sort(u.processes.begin(), u.processes.end(), process_sort_rss());

	while (bytes >= config.thresh && !u.processes.empty()) {
		process::pointer p = u.processes[0];

		std::string const &comm = p->command();
		std::string const &cm = p->cmdline();

		if (!config.debug)
			kill(p->pid(), SIGKILL);

		std::size_t thissize = std::size_t(p->rss());

		log(str(boost::format("    killed process \"%s\" (pid %d) using %dM, usage now %dM")
				% comm % p->pid()
				% (thissize / 1024 / 1024)
				% ((bytes - thissize) / 1024 / 1024)));

		process_list += str(boost::format("    %s (pid %d), using %d megabyte(s).\n"
						  "       command: %s\n")
				% comm % p->pid() % (thissize / 1024 / 1024) % p->cmdline());

		bytes -= thissize;
		u.processes.erase(u.processes.begin());
	}

	log(str(boost::format("    usage is now within acceptable limits (%dM)")
			% (bytes / 1024 / 1024)));

	char nodename[255];
	gethostname(nodename, sizeof nodename);

	std::map<std::string, std::string> msgvars;
	msgvars["limit"] = boost::lexical_cast<std::string>(config.limit / 1024 / 1024);
	msgvars["thresh"] = boost::lexical_cast<std::string>(config.thresh / 1024 / 1024);
	msgvars["user"] = uname;
	msgvars["hostname"] = nodename;
	msgvars["current"] = boost::lexical_cast<std::string>(bytes / 1024 / 1024);
	msgvars["processes"] = process_list;

	std::string message = replace_file(config.mailmessage, msgvars);
	if (!config.debug && !message.empty())
		sendmail(uname, message);
}
